### ISSUE TITLE
Enable getting graph data in a description set

### ISSUE THREAD

**Issue by siuc-nate** [CredentialEngine/CredentialRegistry#475]
We have a potential use case where we want to get:
1. A resource
2. The description set (as in, the connections/URIs) for that resource
3. The graph data for that resource (as in, everything you'd get if you retrieved the resource via its `/graph/` URI)
4. Optionally, related items up to the per branch limit

In other words, the same functionality we have now for description sets, but with the twist of being able to guarantee that everything from that resource's `@graph` (e.g. competencies for frameworks, bnodes for credentials, etc) is included regardless of the other parameters.

This would allow us to get, for example, a Concept Scheme and all of its Concepts, along with the URIs for the first `n` things that reference those concepts, but not waste bandwidth actually getting those extra things. This would be useful for a concept scheme that is referenced by a lot of credentials.  

It would similarly be useful for getting a large and popular Competency Framework and all of its Competencies but not getting (potentially) hundreds of credentials along with it, while still being able to know what those credentials are via the description set's lists of URIs.

Today, this would be accomplished by:
1. Getting the `/graph/` data for the resource
2. Getting the description set data for the resource (optionally including related items up to the per branch limit)
3. Injecting all of the stuff from the resource's `@graph` (except for the resource itself, since that's already returned in a separate field) into the description set's "related items" array (minus any that are already in there)
4. Returning the combined result

Am I making sense? If so, what is the feasibility of adding such a feature? Ideally it would just be another parameter that could be added to the existing search/description set stuff, perhaps something like `include_graph_data` or some similar parameter.

It's not needed urgently, but it would be nice to have.

**Comment by excelsior** [comment]
@siuc-nate This improvement is available on sandbox and production.

Setting `include_graph_data` to `true` will return all of the resources from the `@graph`s of the envelopes with given CTIDs (except the top-level resources) under the `description_set_resources` property.

**Comment by siuc-nate** [comment]
Thanks, it works nicely.

It looks like I was mistaken when I initially mentioned the resource itself, specifically when I said:
>since that's already returned in a separate field

Turns out it's not included in the DSP data, which is something I had forgotten about. I was fetching it separately but I hadn't looked at that part of the code in a while, so I didn't notice that it wasn't coming directly from the description set response data.

I hate to bug you one more time about this, but, would it be possible to also:
- Return the resource(s) for each CTID in a `data` object (just as we do with the search results)
- Ensure that the metadata (if requested) is provided for the resource(s) from the `ctids` list in the initial request

The end result of this would be that a request like this:
```
{
    "ctids": [
		"ce-62397436-4712-4af5-92e6-74cc7ac2b00a"
	],
	"include_resources": true,
	"include_graph_data": true,
	"per_branch_limit": 2,
	"include_results_metadata": true
}
```
Would return a response like this:
```
{
    "data": [
        { "...": "the resource data for ce-62397436-4712-4af5-92e6-74cc7ac2b00a" }
    ],
    "description_sets": [
        {
            "ctid": "ce-62397436-4712-4af5-92e6-74cc7ac2b00a",
            "...": "other data, as is currently being returned"
        }
    ],
    "description_set_resources": [
        { "...": "The combined graph data and other related resources, as are currently being returned" }
    ],
    "results_metadata": [
        {
            "resource_uri": "ce-62397436-4712-4af5-92e6-74cc7ac2b00a",
            "...": "other data, as is currently being returned"
        }
    ]
}
```
It's probably not necessary to return the metadata for the related resources, just for those specified in the `ctids` array in the request.  If we need the metadata for the related stuff, that should probably be controlled by a separate parameter, but I don't have a use case for that right now so I wouldn't worry about it.

Anyway, this change would enable getting everything with just one request:
- The resource(s) specified in the `ctids` array
- The DSP URIs for the specified resource(s) (up to the `per_branch_limit`)
- The related data the specified resource(s) (up to the `per_branch_limit`)
- The graph data for the specified resource(s) (regardless of the `per_branch_limit`)
- The metadata for the specified resource(s)

We seem to have a lot of overlapping functionality in terms of ways to get data (via search, via CTID, etc) and it might be useful to try to compile all of the options in one place and see if there are any gaps in functionality that need to be filled and/or places where the code can be simplified because it's doing the same things in different spots.

**Comment by siuc-nate** [comment]
I'm liking it so far. Can we enable this on production as well?

Thanks.

**Comment by siuc-nate** [comment]
Can this be added to the search API so that something like this query will return the graph data in the same way/structure/format it works above?:

![image](https://user-images.githubusercontent.com/21346519/140779285-f23085a4-6779-490e-9cd3-6f4605b6e13c.png)

We should probably come up with a listing of all the ways there currently are to get data out of the registry and what options each one supports. There is a lot of overlap but there are gaps here and there, and while the formatting is mostly consistent, it would probably be good if it were consistent throughout.

**Comment by excelsior** [comment]
@siuc-nate Done, but the parameter is called `include_description_set_graph_data` for the sake of consistency.

BTW, Steve suggested multiple times to use a GraphQL interface for defining queries. I, too, think this will be a great approach for our case.

**Comment by siuc-nate** [comment]
Thanks.
This is for the search results though - I'm curious what led to the use of "description set" in the parameter name.

**Comment by excelsior** [comment]
Oops, I thought you were interested in fetching description sets' graph data.

The description sets API allows to get subresources (`include_resources`) and graph data (`include_graph_data`) alongside the description sets themselves.

The search API has the `include_description_set_resources` parameter, too, which fetches _description sets'_ subresources. My understanding was the other parameter should enable getting description sets' graph data, too. Hence the name change.

If it's all about the search results, how should the response look like?

**Comment by siuc-nate** [comment]
I think part of the confusion is/was our use of `include_graph_data` as the parameter but `description_set_resources` as the name of the property in the response.

To clarify, what I meant was, can we pass `include_graph_data=true` as part of a search request and receive the graph data for the search results inside of the same structure used elsewhere (meaning inside of the `description_set_resources` array)?

That way it can work the same way as it does for the description set API where there are varying options as to how much (if any) description set data is returned but we can ensure that the graph data for the search results is always returned regardless of any other description set parameters/data as long as `include_graph_data` is true. I guess the response property name (`description_set_resources`) is a little unintuitive if the query doesn't actually ask for any description set data, but if it does ask for it, then it makes sense, so I'm fine with using that property in either case.

In other words:
Request:
```
https://credentialengineregistry.org/ctdl?skip=0&take=5&include_graph_data=true
{
  "@type": "ceasn:CompetencyFramework"
}
```

Response:
```
{
    "data": [
        { "...": "the search results data, as currently returned today" }
    ],
    "description_set_resources": [
        { "...": "The combined graph data for the search results" }
    ]
}
```

If being able to get the graph data back for a search is already available some other way that I have completely forgotten about, let me know.

**Comment by excelsior** [comment]
Yep, it's a bit confusing. I actually recall misinterpreting the previous task the same way as well. Anyway, it's fixed now, and the parameter is called `include_graph_data`.

**Comment by siuc-nate** [comment]
Thanks, it looks like it's working.

However, it also seems to be returning the `description_sets` array regardless of whether or not `include_description_sets` is set to true. Is this intentional? If all the user needs is the search results + graph data, then returning the description set data is more than is needed.
![image](https://user-images.githubusercontent.com/21346519/142452570-3035ac84-7b64-49b4-8a82-b90f05f1572e.png)

**Comment by excelsior** [comment]
@siuc-nate That bug is fixed now, please take a look.

**Comment by siuc-nate** [comment]
Great, thank you.
